File: supervent.py
import json
import random
import time
from datetime import datetime, timezone
import uuid
import asyncio
import aiohttp
import math
import yaml
import argparse
import signal
import sys

DEFAULT_BATCH_SIZE = 100

class EventGenerator:
    def __init__(self, dataset, api_key, batch_size=DEFAULT_BATCH_SIZE):
        self.dataset = dataset
        self.api_key = api_key
        self.url = f"https://api.axiom.co/v1/datasets/{dataset}/ingest"
        self.batch_size = batch_size
        self.batch = []

    async def emit(self, record):
        # Strip "custom." prefix from keys
        stripped_record = {k.replace("custom_", ""): v for k, v in record.items()}
        # Add a timestamp to the record
        stripped_record['_time'] = datetime.now(timezone.utc).isoformat()

        self.batch.append(stripped_record)
        if len(self.batch) >= self.batch_size:
            await self.send_batch()

    async def send_batch(self):
        if not self.batch:
            return
        # print("sending batch")
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}"
        }
        async with aiohttp.ClientSession() as session:
            async with session.post(self.url, headers=headers, json=self.batch) as response:
                if response.status != 200:
                    print(f"Failed to send batch: {response.status}")
            # else:
                # print("Batch sent successfully")
        self.batch = []

# Load configuration from config.json
def load_config(file_path):
    with open(file_path, 'r') as file:
        config = json.load(file)
    return config

# Load AXIOM_DATASET, AXIOM_API_KEY, and HTTP_BATCH_SIZE from axiom_config.yaml
def load_axiom_config(file_path):
    with open(file_path, 'r') as file:
        config = yaml.safe_load(file)
    return config

# Generate a random event based on the source configuration
def generate_event(source_config):
    event = {"source": source_config["vendor"]}
    for field, details in source_config['fields'].items():
        if details['type'] == 'datetime':
            if source_config['timestamp_format'] == 'UTC':
                event[field] = datetime.utcnow().strftime(details.get('format', '%Y-%m-%dT%H:%M:%SZ'))
            elif source_config['timestamp_format'] == 'ISO':
                event[field] = datetime.now().isoformat()
            elif source_config['timestamp_format'] == 'Unix':
                event[field] = int(time.time())
            elif source_config['timestamp_format'] == 'RFC3339':
                event[field] = datetime.now().strftime('%Y-%m-%dT%H:%M:%SZ')
            else:
                event[field] = datetime.now().strftime(source_config['timestamp_format'])
        elif details['type'] == 'string':
            if 'constraints' in details and 'allowed_values' in details['constraints']:
                event[field] = random.choice(details['constraints']['allowed_values'])
            else:
                if field == "message":
                    if 'formats' in details:
                        selected_format = random.choice(details['formats'])
                        event[field] = selected_format.format(
                            timestamp=datetime.now().strftime(details.get('format', '%Y-%m-%dT%H:%M:%SZ')),
                            src_ip=generate_random_ip_address() if '{src_ip}' in selected_format else '',
                            dst_ip=generate_random_ip_address() if '{dst_ip}' in selected_format else '',
                            ip_address=generate_random_ip_address() if '{ip_address}' in selected_format else ''
                        )
                    else:
                        event[field] = details['format'].format(
                            timestamp=datetime.now().strftime(details.get('format', '%Y-%m-%dT%H:%M:%SZ')),
                            src_ip=generate_random_ip_address() if '{src_ip}' in details['format'] else '',
                            dst_ip=generate_random_ip_address() if '{dst_ip}' in details['format'] else '',
                            ip_address=generate_random_ip_address() if '{ip_address}' in details['format'] else ''
                        )
                else:
                    event[field] = generate_random_username() if field == "user" else generate_random_ip_address()
        elif details['type'] == 'int':
            min_val = int(details.get('constraints', {}).get('min', 0))
            max_val = int(details.get('constraints', {}).get('max', 100))
            event[field] = random.randint(min_val, max_val)
        # Add more types as needed
    return event

# Generate a random IP address
def generate_random_ip_address():
    return f"{random.randint(1, 255)}.{random.randint(0, 255)}.{random.randint(0, 255)}.{random.randint(1, 255)}"

# Generate a random username based on Zipf's law
def generate_random_username():
    usernames = [
        "john_doe", "jane_smith", "mohamed_ali", "li_wei", "maria_garcia",
        "yuki_tanaka", "olga_petrov", "raj_kumar", "fatima_zahra", "chen_wang",
        "ahmed_hassan", "isabella_rossi", "david_jones", "sophia_martinez", "emily_clark",
        "noah_brown", "mia_wilson", "lucas_miller", "oliver_davis", "ava_moore",
        "ethan_taylor", "amelia_anderson", "james_thomas", "harper_jackson", "benjamin_white",
        "liam_johnson", "emma_rodriguez", "william_lee", "sophia_kim", "mason_martin",
        "elijah_hernandez", "logan_lopez", "alexander_gonzalez", "sebastian_perez", "daniel_hall",
        "matthew_young", "henry_king", "jack_wright", "levi_scott", "isaac_green",
        "gabriel_baker", "julian_adams", "jayden_nelson", "lucas_carter", "anthony_mitchell",
        "grayson_perez", "dylan_roberts", "leo_turner", "jaxon_phillips", "asher_campbell",
        "ananya_sharma", "arjun_patel", "priya_singh", "vikram_gupta", "neha_verma",
        "sanjay_rana", "deepika_kapoor", "ravi_mehta", "sara_khan", "manoj_joshi",
        "željko_ivanović", "šime_šarić", "đorđe_đorđević", "čedomir_čolić", "žana_živković",
        "miloš_milošević", "ana_marija", "ivan_ivanov", "petar_petrov", "nikola_nikolić",
        "marta_novak", "katarina_kovač", "tomaž_tomažič", "matej_matejić", "vanja_vuković",
        "dragana_dimitrijević", "bojan_bojović", "milica_milovanović", "stefan_stefanović", "vanja_vasić",
        "igor_ilić", "jelena_jovanović", "marko_marković", "tanja_tomić", "zoran_zorić"
    ]

    # Generate weights using Zipf's law
    weights = [1.0 / (i + 1) for i in range(len(usernames))]
    total_weight = sum(weights)
    normalized_weights = [w / total_weight for w in weights]

    # Select a username based on the weights
    r = random.random()
    cumulative_weight = 0.0
    for username, weight in zip(usernames, normalized_weights):
        cumulative_weight += weight
        if r < cumulative_weight:
            return username
    return usernames[-1]

# Signal handler to gracefully exit on ^C or kill signal
def signal_handler(signal, frame):
    print("Received interrupt signal, sending remaining events...")
    if len(event_generator.batch) > 0:
        asyncio.create_task(event_generator.send_batch())
    sys.exit(0)

# Main function to generate events
async def main():
    parser = argparse.ArgumentParser(description='Generate and send events.')
    parser.add_argument('--config', type=str, default='config.json', help='Path to the configuration file')
    args = parser.parse_args()

    config = load_config(args.config)
    axiom_config = load_axiom_config('axiom_config.yaml')
    dataset = axiom_config['AXIOM_DATASET']
    api_key = axiom_config['AXIOM_API_KEY']
    batch_size = int(axiom_config.get('HTTP_BATCH_SIZE', DEFAULT_BATCH_SIZE))
    global event_generator
    event_generator = EventGenerator(dataset=dataset, api_key=api_key, batch_size=batch_size)

    # Set up signal handling to gracefully exit on ^C or kill signal
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Generate events in a round-robin fashion indefinitely
    while True:
        for source in config['sources']:
            event = generate_event(source)
            await event_generator.emit(event)

if __name__ == "__main__":
    asyncio.run(main())

File: supervent.go
package main

import (
	"database/sql"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"os/signal"
	"time"

	"github.com/google/uuid"
	_ "github.com/lib/pq"
	"github.com/valyala/fasthttp"
)

const DEFAULT_BATCH_SIZE = 100

type Config struct {
	Sources []SourceConfig `json:"sources"`
}

type SourceConfig struct {
	Vendor          string           `json:"vendor"`
	TimestampFormat string           `json:"timestamp_format"`
	Fields          map[string]Field `json:"fields"`
}

type Field struct {
	Type          string                 `json:"type"`
	AllowedValues []string               `json:"allowed_values,omitempty"`
	Weights       []float64              `json:"weights,omitempty"`
	Constraints   map[string]interface{} `json:"constraints,omitempty"`
	Distribution  string                 `json:"distribution,omitempty"`
	Mean          float64                `json:"mean,omitempty"`
	Stddev        float64                `json:"stddev,omitempty"`
	Lambda        float64                `json:"lambda,omitempty"`
	S             float64                `json:"s,omitempty"`
	Alpha         float64                `json:"alpha,omitempty"`
	Format        string                 `json:"format,omitempty"`
	Formats       []string               `json:"formats,omitempty"`
}

type EventGenerator struct {
	Dataset      string
	APIKey       string
	URL          string
	BatchSize    int
	Batch        []map[string]interface{}
	PostgresConn *sql.DB
}

func NewEventGenerator(dataset, apiKey string, batchSize int, postgresConfig *PostgresConfig) *EventGenerator {
	eg := &EventGenerator{
		Dataset:   dataset,
		APIKey:    apiKey,
		URL:       fmt.Sprintf("https://api.axiom.co/v1/datasets/%s/ingest", dataset),
		BatchSize: batchSize,
		Batch:     make([]map[string]interface{}, 0, batchSize),
	}

	if postgresConfig != nil {
		connStr := fmt.Sprintf("host=%s port=%d dbname=%s user=%s password=%s sslmode=disable",
			postgresConfig.Host, postgresConfig.Port, postgresConfig.DBName, postgresConfig.User, postgresConfig.Password)
		db, err := sql.Open("postgres", connStr)
		if err != nil {
			log.Fatalf("Failed to connect to PostgreSQL: %v", err)
		}
		eg.PostgresConn = db
	}

	return eg
}

func (eg *EventGenerator) Emit(record map[string]interface{}) {
	record["_time"] = time.Now().UTC().Format(time.RFC3339)
	eg.Batch = append(eg.Batch, record)
	if len(eg.Batch) >= eg.BatchSize {
		eg.SendBatch()
	}
}

func (eg *EventGenerator) SendBatch() {
	if len(eg.Batch) == 0 {
		return
	}

	fmt.Println("sending batch")
	headers := map[string]string{
		"Content-Type":  "application/json",
		"Authorization": fmt.Sprintf("Bearer %s", eg.APIKey),
	}

	body, err := json.Marshal(eg.Batch)
	if err != nil {
		log.Fatalf("Failed to marshal batch: %v", err)
	}

	req := fasthttp.AcquireRequest()
	req.SetRequestURI(eg.URL)
	req.Header.SetMethod("POST")
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	req.SetBody(body)

	resp := fasthttp.AcquireResponse()
	client := &fasthttp.Client{}
	if err := client.Do(req, resp); err != nil {
		log.Printf("Failed to send batch: %v", err)
	} else if resp.StatusCode() != fasthttp.StatusOK {
		log.Printf("Failed to send batch: %d", resp.StatusCode())
	} else {
		fmt.Println("Batch sent successfully")
	}

	if eg.PostgresConn != nil {
		eg.SendToPostgres(eg.Batch)
	}

	eg.Batch = eg.Batch[:0]
	fasthttp.ReleaseRequest(req)
	fasthttp.ReleaseResponse(resp)
}

func (eg *EventGenerator) SendToPostgres(batch []map[string]interface{}) {
	for _, record := range batch {
		columns := make([]string, 0, len(record))
		values := make([]interface{}, 0, len(record))
		for k, v := range record {
			columns = append(columns, k)
			values = append(values, v)
		}
		insertStatement := fmt.Sprintf("INSERT INTO %s (%s) VALUES (%s)",
			eg.Dataset, join(columns, ","), placeholders(len(values)))
		_, err := eg.PostgresConn.Exec(insertStatement, values...)
		if err != nil {
			log.Printf("Failed to insert record into PostgreSQL: %v", err)
		}
	}
}

func join(strs []string, sep string) string {
	result := ""
	for i, s := range strs {
		if i > 0 {
			result += sep
		}
		result += s
	}
	return result
}

func placeholders(n int) string {
	result := ""
	for i := 0; i < n; i++ {
		if i > 0 {
			result += ","
		}
		result += fmt.Sprintf("$%d", i+1)
	}
	return result
}

type PostgresConfig struct {
	Host     string
	Port     int
	DBName   string
	User     string
	Password string
}

func loadConfig(filePath string) (*Config, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var config Config
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&config); err != nil {
		return nil, err
	}

	return &config, nil
}

func generateEvent(sourceConfig SourceConfig) map[string]interface{} {
	event := map[string]interface{}{
		"source": sourceConfig.Vendor,
	}
	for field, details := range sourceConfig.Fields {
		switch details.Type {
		case "datetime":
			event[field] = generateDatetime(details, sourceConfig.TimestampFormat)
		case "string":
			event[field] = generateString(details, field)
		case "int":
			event[field] = generateInt(details)
		}
	}
	fmt.Println(event) // Debug statement to print the complete event
	return event
}

func generateDatetime(details Field, timestampFormat string) string {
	switch timestampFormat {
	case "UTC":
		return time.Now().UTC().Format(details.Format)
	case "ISO":
		return time.Now().Format(time.RFC3339)
	case "Unix":
		return fmt.Sprintf("%d", time.Now().Unix())
	case "RFC3339":
		return time.Now().Format(time.RFC3339)
	default:
		return time.Now().Format(details.Format)
	}
}

func generateString(details Field, fieldName string) string {
	if len(details.AllowedValues) > 0 {
		if len(details.Weights) > 0 {
			return weightedChoice(details.AllowedValues, details.Weights)
		}
		return details.AllowedValues[rand.Intn(len(details.AllowedValues))]
	}
	if len(details.Formats) > 0 {
		selectedFormat := details.Formats[rand.Intn(len(details.Formats))]
		return fmt.Sprintf(selectedFormat, time.Now().Format(details.Format))
	}
	if fieldName == "client_ip" || fieldName == "server_ip" || fieldName == "source_ip" || fieldName == "destination_ip" {
		return generateRandomIPAddress()
	}
	return uuid.New().String()
}

func generateInt(details Field) int {
	min := int(details.Constraints["min"].(float64))
	max := int(details.Constraints["max"].(float64))
	switch details.Distribution {
	case "uniform":
		return rand.Intn(max-min+1) + min
	case "normal":
		return int(rand.NormFloat64()*details.Stddev + details.Mean)
	case "exponential":
		return int(rand.ExpFloat64() / details.Lambda)
	case "zipfian":
		return int(randZipf(details.S))
	case "long_tail":
		return int(randPareto(details.Alpha))
	case "random":
		return rand.Intn(max-min+1) + min
	default:
		return rand.Intn(max-min+1) + min
	}
}

func generateRandomIPAddress() string {
	return fmt.Sprintf("%d.%d.%d.%d", rand.Intn(256), rand.Intn(256), rand.Intn(256), rand.Intn(256))
}

func weightedChoice(values []string, weights []float64) string {
	totalWeight := 0.0
	for _, weight := range weights {
		totalWeight += weight
	}
	r := rand.Float64() * totalWeight
	for i, weight := range weights {
		if r < weight {
			return values[i]
		}
		r -= weight
	}
	return values[len(values)-1]
}

func randZipf(s float64) float64 {
	return rand.ExpFloat64() / s
}

func randPareto(alpha float64) float64 {
	return rand.ExpFloat64() / alpha
}

func main() {
	configPath := flag.String("config", "config.json", "Path to the configuration file")
	dataset := flag.String("dataset", "", "Axiom dataset name")
	apiKey := flag.String("api_key", "", "Axiom API key")
	batchSize := flag.Int("batch_size", DEFAULT_BATCH_SIZE, "Batch size for HTTP requests")
	postgresHost := flag.String("postgres_host", "", "PostgreSQL host")
	postgresPort := flag.Int("postgres_port", 5432, "PostgreSQL port")
	postgresDB := flag.String("postgres_db", "", "PostgreSQL database name")
	postgresUser := flag.String("postgres_user", "", "PostgreSQL user")
	postgresPassword := flag.String("postgres_password", "", "PostgreSQL password")
	flag.Parse()

	if *dataset == "" || *apiKey == "" {
		log.Fatal("dataset and api_key are required")
	}

	config, err := loadConfig(*configPath)
	if err != nil {
		log.Fatalf("Failed to load config: %v", err)
	}

	var postgresConfig *PostgresConfig
	if *postgresHost != "" && *postgresDB != "" && *postgresUser != "" && *postgresPassword != "" {
		postgresConfig = &PostgresConfig{
			Host:     *postgresHost,
			Port:     *postgresPort,
			DBName:   *postgresDB,
			User:     *postgresUser,
			Password: *postgresPassword,
		}
	}

	eventGenerator := NewEventGenerator(*dataset, *apiKey, *batchSize, postgresConfig)

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, os.Kill)
	go func() {
		<-sigChan
		fmt.Println("Received interrupt signal, sending remaining events...")
		eventGenerator.SendBatch()
		if eventGenerator.PostgresConn != nil {
			eventGenerator.PostgresConn.Close()
		}
		os.Exit(0)
	}()

	for {
		for _, source := range config.Sources {
			event := generateEvent(source)
			eventGenerator.Emit(event)
		}
		time.Sleep(1 * time.Second) // Adjust the sleep duration as needed
	}
}
